### ISSUE TITLE
Bound and supertrait bounds can be empty

### ISSUE THREAD

**Issue by bluss** [rust-lang/rust#30106]
These examples all compile:

``` rust
trait A : { }

struct Foo<T: > { t: T }

fn foo<T: >() { }
```

I expected empty bounds lists to be an error.

**Comment by Gankro** [comment]
I could see this being useful for macros?

**Comment by eddyb** [comment]
cc @petrochenkov

**Comment by petrochenkov** [comment]
It was decided this is valid syntax, see https://github.com/rust-lang/rust/pull/37511 and https://github.com/rust-lang/rust/pull/39158.
